controllers/constraintstatus_controller.go

```diff
@@ -0,0 +1,418 @@
+	*[]v1alpha1.SyncOnlyEntry, error,
+) {
+	syncOnlys := []v1alpha1.SyncOnlyEntry{}
+
+	for _, match := range constraintMatchKinds {
+		newKind, ok := match.(map[string]interface{})
+		if !ok {
+			continue
+		}
+
+		apiGroups, ok := newKind["apiGroups"].([]interface{})
+		if !ok {
+			continue
+		}
+
+		kindsInKinds, ok := newKind["kinds"].([]interface{})
+		if !ok {
+			continue
+		}
+
+		for _, apiGroup := range apiGroups {
+			for _, kind := range kindsInKinds {
+				version, err := r.getAPIVersion(kind.(string), apiGroup.(string), false)
+				if err != nil {
+					return nil, err
+				}
+
+				syncOnlys = append(syncOnlys, v1alpha1.SyncOnlyEntry{
+					Group:   apiGroup.(string),
+					Version: version,
+					Kind:    kind.(string),
+				})
+			}
+		}
+	}
+
+	return &syncOnlys, nil
+}
+
+func (r *ConstraintStatusReconciler) getAPIVersion(kind string, apiGroup string, skipRefresh bool) (string, error) {
+	// cool time(10 min) to refresh discoveries
+	if len(r.apiResourceList) == 0 ||
+		r.discoveryLastRefreshed.Add(time.Minute*10).Before(time.Now()) {
+		err := r.refreshDiscoveryInfo()
+		r.discoveryLastRefreshed = time.Now()
+
+		if err != nil {
+			return "", err
+		}
+	}
+
+	for _, resc := range r.apiResourceList {
+		groupVerison, err := schema.ParseGroupVersion(resc.GroupVersion)
+		if err != nil {
+			r.Log.Error(err, "Cannot Parse Group and version in getApiVersion ", "GroupVersion:", resc.GroupVersion)
+
+			continue
+		}
+
+		group := groupVerison.Group
+		version := groupVerison.Version
+		// Consider groupversion == v1 or groupversion == app1/v1
+		for _, apiResource := range resc.APIResources {
+			if apiResource.Kind == kind && group == apiGroup {
+				return version, nil
+			}
+		}
+	}
+
+	if !skipRefresh {
+		// Get new discoveryInfo, when any resource is not found
+		err := r.refreshDiscoveryInfo()
+		if err != nil {
+			return "", err
+		}
+
+		r.discoveryLastRefreshed = time.Now()
+		// Retry one more time after refresh the discovery
+		return r.getAPIVersion(kind, apiGroup, true)
+	}
+
+	return "", ErrNotFoundDiscovery
+}
+
+// Retrieve all groups and versions to add in config sync
+// Constraints present only kind and group so this function helps to find the version
+func (r *ConstraintStatusReconciler) refreshDiscoveryInfo() error {
+	discoveryClient := r.ClientSet.Discovery()
+
+	apiList, err := discoveryClient.ServerPreferredResources()
+	if err != nil {
+		return err
+	}
+
+	r.apiResourceList = apiList
+
+	return nil
+}
+
+func (r *ConstraintStatusReconciler) getUniqSyncOnly() []v1alpha1.SyncOnlyEntry {
+	syncOnlySet := map[v1alpha1.SyncOnlyEntry]bool{}
+	// Add to table for unique filtering
+	r.constraintMap.Range(func(ckey interface{}, entries interface{}) bool {
+		constraintName, ok := ckey.(string)
+		if !ok {
+			r.Log.V(1).Info("Error to parse constraint name")
+
+			return true
+		}
+
+		arrayEntries, ok := entries.([]v1alpha1.SyncOnlyEntry)
+		if !ok {
```

Comment:
I think you can skip the type assert `ok` check since the value is guaranteed to always be this type since this controller is the only thing that populates it.